source/adapters/level_zero/adapter.cpp

```diff
@@ -65,24 +90,37 @@ ur_result_t initPlatforms(PlatformVec &platforms) noexcept try {
 
   ZE2UR_CALL(zeDriverGet, (&ZeDriverCount, ZeDrivers.data()));
   for (uint32_t I = 0; I < ZeDriverCount; ++I) {
+    bool DriverInit = false;
     ze_device_properties_t device_properties{};
     device_properties.stype = ZE_STRUCTURE_TYPE_DEVICE_PROPERTIES;
     uint32_t ZeDeviceCount = 0;
     ZE2UR_CALL(zeDeviceGet, (ZeDrivers[I], &ZeDeviceCount, nullptr));
     ZeDevices.resize(ZeDeviceCount);
     ZE2UR_CALL(zeDeviceGet, (ZeDrivers[I], &ZeDeviceCount, ZeDevices.data()));
+    auto platform = std::make_unique<ur_platform_handle_t_>(ZeDrivers[I]);
     // Check if this driver has GPU Devices
     for (uint32_t D = 0; D < ZeDeviceCount; ++D) {
       ZE2UR_CALL(zeDeviceGetProperties, (ZeDevices[D], &device_properties));
-
       if (ZE_DEVICE_TYPE_GPU == device_properties.type) {
-        // If this Driver is a GPU, save it as a usable platform.
-        auto platform = std::make_unique<ur_platform_handle_t_>(ZeDrivers[I]);
-        UR_CALL(platform->initialize());
+        if (!DriverInit) {
+          // If this Driver is a GPU, save it as a usable platform.
+          UR_CALL(platform->initialize());
 
-        // Save a copy in the cache for future uses.
-        platforms.push_back(std::move(platform));
-        break;
+          // Save a copy in the cache for future uses.
+          platforms.push_back(std::move(platform));
+          DriverInit = true;
+        }
+        if (ZesResult == ZE_RESULT_SUCCESS) {
+          ur_zes_device_handle_data_t ZesDeviceData;
+          zes_uuid_t ZesUUID;
+          std::memcpy(&ZesUUID, &device_properties.uuid, sizeof(zes_uuid_t));
+          if (getZesDeviceHandle(
+                  ZesUUID, &ZesDeviceData.ZesDevice, &ZesDeviceData.SubDeviceId,
+                  &ZesDeviceData.SubDevice) == UR_RESULT_SUCCESS) {
+            platforms.back()->ZedeviceToZesDeviceMap.insert(
```

Comment:
I don't think this is right. This code should move into the platform->initialize() function. Right now you will add all the mappings to the last platform in the container.

Reply:
This is adding the device mappings to the last platform that was added to the platforms container which matches the current driver whose devices are being examined.

The next time in the loop for the next driver, we would be adding the mappings to the next last platform added.

This is supposed to be adding the map to the last platform that has been added each time after then platform has been pushed back.

If you feel strongly about adding this to the init, I can see if that is possible, but it would require going back thru the zedevices again. The reason I have this here is because we are already going thru the zedevices, then this calls the function to further loop thru the zesdevices to find the match.

Reply:
I updated the variable names used and added comments to explain the process. Please let me know if you still feel the logic is incorrect.

Reply:
ah, ok, I see, the two outer loops threw me off :) I still think I'd be cleaner to have this inside of platform initialize, but that's less important.  No need to change it.